analyses/stackanalyses/controller.go

```diff
@@ -0,0 +1,191 @@
+		Factor: 2,
+		Jitter: false,
+	}
+	matcher, err := GetMatcher(requestParams.RawManifestFile)
+	if err != nil {
+		log.Fatal().Msgf(err.Error())
+	}
+	mc := NewController(matcher)
+	mc.fileStats = mc.buildFileStats(requestParams.RawManifestFile)
+	postResponse := mc.postRequest(requestParams, mc.fileStats.DepsTreePath)
+	getResponse := mc.getRequest(requestParams, postResponse, b)
+	log.Debug().Msgf("Success StackAnalyses.")
+	return getResponse
+}
+
+// GetManifestFilePath sets file path
+func (mc *Controller) GetManifestFilePath(input string) string {
+	path, err := filepath.Abs(input)
+	if err != nil {
+		log.Fatal().Msgf("Invalid Path of Manifest file. Only Absolute path is allowed.")
+	}
+	return path
+}
+
+// postRequest performs Stack Analyses POST Request to CRDA server.
+func (mc *Controller) postRequest(requestParams driver.RequestType, filePath string) driver.PostResponseType {
+	log.Debug().Msgf("Executing: postRequest.")
+	manifest := &bytes.Buffer{}
+	requestData := utils.HTTPRequestType{
+		Method:          http.MethodPost,
+		Endpoint:        APIStackAnalyses,
+		ThreeScaleToken: requestParams.ThreeScaleToken,
+		Host:            requestParams.Host,
+	}
+	writer := multipart.NewWriter(manifest)
+	fd, err := os.Open(filePath)
+	if err != nil {
+		log.Fatal().Err(err).Msgf(err.Error())
+	}
+	defer fd.Close()
+
+	fw, err := writer.CreateFormFile("manifest", mc.m.DepsTreeFileName())
+	if err != nil {
+		log.Fatal().Err(err).Msgf(err.Error())
+	}
+	_, err = io.Copy(fw, fd)
+	if err != nil {
+		log.Fatal().Err(err).Msgf(err.Error())
+	}
+	_ = writer.WriteField("ecosystem", mc.m.Ecosystem())
+	_ = writer.WriteField("file_path", "/tmp/bin")
+	err = writer.Close()
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Error closing Buffer Writer in Stack Analyses Request.")
+	}
+	log.Debug().Msgf("Hitting: Stack Analyses Post API.")
+	apiResponse := utils.HTTPRequestMultipart(requestData, writer, manifest)
+	body := mc.validatePostResponse(apiResponse)
+	log.Debug().Msgf("Got Stack Analyses Post Response Stack Id: %s", body.ID)
+	log.Debug().Msgf("Success: postRequest.")
+	return body
+}
+
+// getRequest performs Stack Analyses GET Request to CRDA Server.
+func (mc *Controller) getRequest(requestParams driver.RequestType, postResponse driver.PostResponseType, back *backoff.Backoff) driver.GetResponseType {
+	log.Debug().Msgf("Executing: getRequest.")
+	requestData := utils.HTTPRequestType{
+		Method:          http.MethodGet,
+		Endpoint:        APIStackAnalyses + "/" + postResponse.ID,
+		ThreeScaleToken: requestParams.ThreeScaleToken,
+		Host:            requestParams.Host,
+	}
+	d := back.Duration()
+	log.Debug().Msgf("Sleeping for %s", d)
+	time.Sleep(d)
+	apiResponse := utils.HTTPRequest(requestData)
+	if apiResponse.StatusCode == http.StatusAccepted {
+		// Retry till server respond 200 or Timeout Error or Exponential Backoff limit hit.
+		log.Debug().Msgf("Retrying...")
+		mc.getRequest(requestParams, postResponse, back)
```

Comment:
Why not a loop instead of recursion to achieve exponential backoff?

Even the sample code from backup is using loop.
https://github.com/jpillora/backoff#example-using-net-package

Reply:
It should be okay either way.

Reply:
Update to loop